src/hnsw/failure_point.c

```diff
@@ -0,0 +1,75 @@
+#include <postgres.h>
+
+#include "hnsw/failure_point.h"
+
+#include <inttypes.h> /* PRIu32 */
+
+struct failure_point_state
+{
+    bool        enabled;
+    const char *func;
+    const char *name;
+    uint32      remaining;
+};
+
+static struct failure_point_state *failure_point_get_state(void)
+{
+    static struct failure_point_state state = {};
+
+    return &state;
+}
+
+void ldb_failure_point_enable(const char *func, const char *name, uint32 dont_trigger_first_nr)
+{
+    struct failure_point_state *state = failure_point_get_state();
+
+    if(!LANTERN_FAILURE_POINTS_ARE_ENABLED) {
+        elog(WARNING,
+             "Can't enable failure point for (func=%s name=%s), "
+             "because failure points are disabled in compile time.",
+             func,
+             name);
+    }
+    if(state->enabled) {
+        elog(WARNING,
+             "ldb_failure_point_enable(): another failure point is enabled already."
+             " old failure point: func=%s name=%s remaining=%" PRIu32
+             " new failure point: func=%s name=%s dont_trigger_first_nr=%" PRIu32,
+             state->func,
+             state->name,
+             state->remaining,
+             func,
+             name,
+             dont_trigger_first_nr);
+    }
+    *state = (struct failure_point_state){
+        .enabled = true,
+        .func = func,
+        .name = name,
```

Comment:
Do `name` and `func` live long enough for when it is used from `state`?

Both  go out of scope and are freed as soon as the SQL function returns. It is allocated by the function context. 

so, at that point the global `state.name` is dangling.
If this is the case, you should probably copy the strings. We do not care about performance for these, it should be fine

If you need the name after the

Reply:
Good catch. Done.

I had a choice between using `malloc()` directly, using `char func[0x100];` for the `state->func` and using a custom PostgreSQL allocator for this file. `char func[0x100];` won.